kotlinx-coroutines-core/common/src/channels/ArrayChannel.kt

```diff
@@ -29,10 +30,14 @@ internal open class ArrayChannel<E>(
     }
 
     private val lock = ReentrantLock()
-    private val buffer: Array<Any?> = arrayOfNulls<Any?>(capacity)
+    /*
+     * Guarded by lock.
+     * Allocate minimum of capacity and 16 to avoid excess memory pressure for large channels when it's not necessary.
+     */
+    private var buffer: Array<Any?> = arrayOfNulls<Any?>(min(capacity, 16))
```

Comment:
Live hack: Default allocation size can be made even smaller (I suggest the value of `8` -- it will still produce "quite large" arrays). This way, all the existing stress tests that test for array channels of capacity `10` will also be testing resizing. Unfortunately, they do not actually fail with this buggy code, so a dedicated resizing test is still needed (need to send/receive a bunch a values to wrap around the buffer, then increase the number of buffered values to resize wrapped buffer)